src/Form/Extension/CartTypeExtension.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of Monsieur Biz' Advanced Promotion plugin for Sylius.
+ * (c) Monsieur Biz <sylius@monsieurbiz.com>
+ * For the full copyright and license information, please view the LICENSE.txt
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace MonsieurBiz\SyliusAdvancedPromotionPlugin\Form\Extension;
+
+use MonsieurBiz\SyliusAdvancedPromotionPlugin\Entity\PromotionCouponsAwareInterface;
+use Sylius\Bundle\OrderBundle\Form\Type\CartType;
+use Sylius\Bundle\PromotionBundle\Form\Type\PromotionCouponToCodeType;
+use Sylius\Component\Core\Model\PromotionCouponInterface;
+use Symfony\Component\Form\AbstractTypeExtension;
+use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Validator\Constraints as Assert;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+
+final class CartTypeExtension extends AbstractTypeExtension
+{
+    /**
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $builder
+            ->add('promotionCoupons', CollectionType::class, [
+                'entry_type' => PromotionCouponToCodeType::class,
+                'entry_options' => [
+                    'attr' => [
+                        'form' => 'sylius_cart',
+                    ],
+                    'constraints' => [
+                        new Assert\Callback(
+                            [$this, 'validatePromotionEntry'],
+                            ['monsieurbiz_advanced_promotion_coupon']
+                        ),
+                    ],
+                ],
+                'required' => false,
+                'allow_add' => true,
+                'allow_delete' => true,
+                'button_add_label' => 'monsieurbiz_sylius_advanced_promotion.coupons.add_coupon',
+                'attr' => [
+                    'class' => 'monsieurbiz-coupons',
+                ],
+            ])
+            ->addEventListener(FormEvents::PRE_SUBMIT, [$this, 'removeEmptyCouponsOnPreSubmit'])
+        ;
+    }
+
+    public function validatePromotionEntry(?PromotionCouponInterface $entry, ExecutionContextInterface $context): void
+    {
+        if (null !== $entry) {
+            return;
+        }
+
+        $context
+                ->buildViolation('sylius.promotion_coupon.is_invalid')
+                ->addViolation()
+        ;
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🤓 nitpick:** ‏fix indentation

```suggestion
        $context
            ->buildViolation('sylius.promotion_coupon.is_invalid')
            ->addViolation()
        ;
```